File: Day156_2025/Q156_GFG.cpp
class Solution {
  public:
    int countPaths(vector<vector<int>>& edges, int V, int src, int dest) {
        // Q. Count the paths
        // Code here
        vector<vector<int>> adj(V);
        vector<int> inDeg(V, 0);
        for (auto e: edges) {
            adj[e[0]].push_back(e[1]);
            inDeg[e[1]]++;
        }
    
        queue<int> q;
        for (int i = 0; i < V; i++) {
            if (inDeg[i] == 0) q.push(i);
        }
    
        vector<int> paths(V, 0);
        paths[src] = 1;
    
        while (!q.empty()) {
            int u = q.front(); q.pop();
            for (int v : adj[u]) {
                paths[v] += paths[u];
                if (--inDeg[v] == 0) {
                    q.push(v);
                }
            }
        }
    
        return paths[dest];
    }
};

File: Day156_2025/Q156_GFG.py
from collections import deque, defaultdict

class Solution:
    def countPaths(self, edges, V, src, dest):
        adj = [[] for _ in range(V)]
        inDeg = [0] * V

        for u, v in edges:
            adj[u].append(v)
            inDeg[v] += 1

        q = deque(i for i in range(V) if inDeg[i] == 0)
        paths = [0] * V
        paths[src] = 1

        while q:
            u = q.popleft()
            for v in adj[u]:
                paths[v] += paths[u]
                inDeg[v] -= 1
                if inDeg[v] == 0:
                    q.append(v)

        return paths[dest]
